File: server/app/models/semen_analysis.py
from sqlalchemy import Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.models.sample import Sample


class SemenAnalysis(Base, TimestampMixin):
    __tablename__ = "semen_analyses"

    id: Mapped[int] = mapped_column(
        primary_key=True,
        index=True,
    )

    sample_id: Mapped[int] = mapped_column(
        ForeignKey("samples.id"),
        unique=True,
        nullable=False,
    )

    # -------------------------
    # General
    # -------------------------

    criteria: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    # -------------------------
    # Macroscopic Examination
    # -------------------------

    volume_ml: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    appearance: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    ph: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    viscosity: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    liquefaction_minutes: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    # -------------------------
    # Microscopic Examination
    # -------------------------

    sperm_concentration_million_ml: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    wbc_concentration_million_ml: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    pus_cells: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    debris: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    agglutination: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    # -------------------------
    # Motility
    # -------------------------

    total_motility_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    progressive_motility_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    rapid_progressive_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    slow_progressive_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    non_progressive_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    immotile_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    # -------------------------
    # Morphology
    # -------------------------

    morphology_normal_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    morphology_abnormal_percent: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    # -------------------------
    # Comments
    # -------------------------

    comments: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    sample: Mapped["Sample"] = relationship(
        "Sample",
        back_populates="analysis",
    )

    @property
    def sample_code(self) -> str:
        return self.sample.sample_code

File: client/src/lib/schemas/semen_analysis.ts
import { z } from "zod";

export const semenAnalysisSchema = z.object({
    sample_code: z.string(),

    // -------------------------
    // General
    // -------------------------

    criteria: z.string().min(1).max(30),

    // -------------------------
    // Macroscopic Examination
    // -------------------------

    volume_ml: z.coerce.number().min(0),

    appearance: z.string().min(1).max(30),

    ph: z.coerce.number().min(0).max(14),

    viscosity: z.string().min(1).max(30),

    liquefaction_minutes: z.coerce.number().min(0),

    // -------------------------
    // Microscopic Examination
    // -------------------------

    sperm_concentration_million_ml: z.coerce.number().min(0),

    wbc_concentration_million_ml: z.coerce.number().min(0),

    pus_cells: z.string().min(1).max(20),

    debris: z.string().min(1).max(20),

    agglutination: z.string().min(1).max(20),

    // -------------------------
    // Motility
    // -------------------------

    total_motility_percent: z.coerce.number().min(0).max(100),

    progressive_motility_percent: z.coerce.number().min(0).max(100),

    rapid_progressive_percent: z.coerce.number().min(0).max(100),

    slow_progressive_percent: z.coerce.number().min(0).max(100),

    non_progressive_percent: z.coerce.number().min(0).max(100),

    immotile_percent: z.coerce.number().min(0).max(100),

    // -------------------------
    // Morphology
    // -------------------------

    morphology_normal_percent: z.coerce.number().min(0).max(100),

    morphology_abnormal_percent: z.coerce.number().min(0).max(100),

    // -------------------------
    // Comments
    // -------------------------

    comments: z.string().optional(),
  })


export type SemenAnalysisFormData = z.infer<
  typeof semenAnalysisSchema
>;

export interface SemenAnalysis
  extends SemenAnalysisFormData {
  id: number;
  created_at: string;
  total_sperm_million: number;
  total_motile_sperm_million: number;
  progressive_motile_sperm_million: number;
  morphologically_normal_sperm_million: number;
}